### ISSUE TITLE
Ndt7: Add a way to measure latency

### ISSUE THREAD

**Issue by EKami** [m-lab/ndt-server#322]
Hi everyone,
Is there a way to measure latency on ndt7 servers? So far I can only get results like this one:

```
    {
      "machine": "mlab1-bom02.mlab-oti.measurement-lab.org",
      "location": {
        "city": "Mumbai",
        "country": "IN"
      },
      "urls": {
        "ws:///ndt/v7/download": "ws://ndt-mlab1-bom02.mlab-oti.measurement-lab.org/ndt/v7/download?access_token=eyJhbGciOiJFZERTQSIsImtpZCI6ImxvY2F0ZV8yMDIwMDQwOSJ9.eyJhdWQiOlsibWxhYjEtYm9tMDIubWxhYi1vdGkubWVhc3VyZW1lbnQtbGFiLm9yZyIsIm1sYWIxLmJvbTAyLm1lYXN1cmVtZW50LWxhYi5vcmciXSwiZXhwIjoxNjAxMzU4OTMwLCJpc3MiOiJsb2NhdGUiLCJzdWIiOiJuZHQifQ.-3QOaLvbNaZ3bUD4jDNGqP17FpR5BQp8DuC5mtQ3SZnFyE2HcYbpZSuTh3N-a1Ff9XLOoBECYgE4heZFu5T1BQ",
        "ws:///ndt/v7/upload": "ws://ndt-mlab1-bom02.mlab-oti.measurement-lab.org/ndt/v7/upload?access_token=eyJhbGciOiJFZERTQSIsImtpZCI6ImxvY2F0ZV8yMDIwMDQwOSJ9.eyJhdWQiOlsibWxhYjEtYm9tMDIubWxhYi1vdGkubWVhc3VyZW1lbnQtbGFiLm9yZyIsIm1sYWIxLmJvbTAyLm1lYXN1cmVtZW50LWxhYi5vcmciXSwiZXhwIjoxNjAxMzU4OTMwLCJpc3MiOiJsb2NhdGUiLCJzdWIiOiJuZHQifQ.-3QOaLvbNaZ3bUD4jDNGqP17FpR5BQp8DuC5mtQ3SZnFyE2HcYbpZSuTh3N-a1Ff9XLOoBECYgE4heZFu5T1BQ",
        "wss:///ndt/v7/download": "wss://ndt-mlab1-bom02.mlab-oti.measurement-lab.org/ndt/v7/download?access_token=eyJhbGciOiJFZERTQSIsImtpZCI6ImxvY2F0ZV8yMDIwMDQwOSJ9.eyJhdWQiOlsibWxhYjEtYm9tMDIubWxhYi1vdGkubWVhc3VyZW1lbnQtbGFiLm9yZyIsIm1sYWIxLmJvbTAyLm1lYXN1cmVtZW50LWxhYi5vcmciXSwiZXhwIjoxNjAxMzU4OTMwLCJpc3MiOiJsb2NhdGUiLCJzdWIiOiJuZHQifQ.-3QOaLvbNaZ3bUD4jDNGqP17FpR5BQp8DuC5mtQ3SZnFyE2HcYbpZSuTh3N-a1Ff9XLOoBECYgE4heZFu5T1BQ",
        "wss:///ndt/v7/upload": "wss://ndt-mlab1-bom02.mlab-oti.measurement-lab.org/ndt/v7/upload?access_token=eyJhbGciOiJFZERTQSIsImtpZCI6ImxvY2F0ZV8yMDIwMDQwOSJ9.eyJhdWQiOlsibWxhYjEtYm9tMDIubWxhYi1vdGkubWVhc3VyZW1lbnQtbGFiLm9yZyIsIm1sYWIxLmJvbTAyLm1lYXN1cmVtZW50LWxhYi5vcmciXSwiZXhwIjoxNjAxMzU4OTMwLCJpc3MiOiJsb2NhdGUiLCJzdWIiOiJuZHQifQ.-3QOaLvbNaZ3bUD4jDNGqP17FpR5BQp8DuC5mtQ3SZnFyE2HcYbpZSuTh3N-a1Ff9XLOoBECYgE4heZFu5T1BQ"
      }
    }
```

and while the ICMP protocol works on `mlab1-bom02.mlab-oti.measurement-lab.org` it's not something that I can use in a Javascript client where only HTTP/HTTPS/Websocket is accepted. It'd be great if it can be measured through the WebSocket on the client side since that would allow the client to also measure the jitter. So far I came up with a hacky solution of using the `wss:///ndt/v7/upload` to upload 1 bit of data, wait for the response and measure the time between the two, then divide by 2. It's working "okay" relative to what I get with a `ping` command but sometimes it's completely not accurate. 

Thanks!

**Comment by laiyi-ohlsen** [comment]
@stephen-soltesz

**Comment by bassosimone** [comment]
I have started working on this and have implemented ~50% of it here: https://github.com/bassosimone/ndt-server/tree/feature/roundtrip. It should take 1-2 more weeks for me to finish it 🤞

**Comment by EKami** [comment]
Awesome news! Thank you so much! And how does it work when it comes to deployment exactly once the feature is released? Is it automatically deployed to all your NDT7 servers?

**Comment by bassosimone** [comment]
@EKami if the PR is accepted upstream, I guess it should be released in a few weeks.

**Comment by stephen-soltesz** [comment]
@EKami the method you described for the "hacky solution" using upload will not work the way you want. The messages sent by the server are [randomly timed][1]. So, there is no helpful correspondence between upload start time, time of first message received from the server, and the system to system ping times.

@EKami what is the use-case for this type of latency measurement? Application level latency measurements (as @bassosimone is proposing) may be inherently larger than the equivalent `ping` measurement from the same systems.

[1]: https://github.com/m-lab/ndt-server/blob/8da06dc9f5c757cf0592e1af47f5da5800050dcc/ndt7/measurer/measurer.go#L78

**Comment by EKami** [comment]
@stephen-soltesz 
yeah, I knew there was something fishy about my implementation lol. The use case is simply to be able to run this with a js client since we can't use ICMP. That would also allow me to calculate the jitter which in turn, can tell if the connection is stable or has some micro-disconnections/ is unstable.

**Comment by bassosimone** [comment]
I have written a prototype and am now evaluating it (see the above commits referencing this issue). It will take a little more time to see whether this works in reasonable way in JavaScript. If I see convincing data, then I will turn this prototype into a spec and submit it to the M-Lab core team for evaluation.

**Comment by EKami** [comment]
Awesome news, thanks a lot @bassosimone :D

**Comment by nosnilmot** [comment]
ping! (excuse the pun)

is there any prospect of this feature being implemented? @bassosimone did anything come from your implementatin/evaluation?

I'm also interested in having this feature (application level latency-only test) available.

**Comment by stephen-soltesz** [comment]
FYI: @mattmathis in case you have thoughts.
